HyperQueue integration

Schlevidon opened the pull request:
This PR is meant to track the collaborative development of a feature that will allow the load balancer to use HyperQueue. 

### Current state of the implementation
The basic idea is the following:
1.  Tell HyperQueue to start a model server whenever a new SLURM job is allocated. The server is then treated as a resource, which allows the scheduler to ensure that each request has exclusive access to a model.
2. Whenever the load balancer receives a request a new HyperQueue job, which requests the resource "model", is submitted. The HQ job doesn't do any actual work and instead determines the URL of the model associated with the SLURM job that the HQ job was assigned to. The HQ job then sends back the URL to the load balancer and simply sleeps to ensure that the scheduler doesn't assign any other requests to this model. Meanwhile the load balancer will use the URL to get the response from the actual model via the UM-Bridge interface. Once the computation is finished the load balancer will return the result to the client and cancel the HQ job to let the model accept new requests.

Here is a more detailed description of the current implementation:
1. The entry point is still `test-LoadBalancer.cpp`, which was left unchanged for now. As before we use a manually submitted SLURM job, which will run the model server and write the URL to a file. We then read the URL, connect to the server to determine which models are available and start a load balancer for every model.
2. On initialization a load balancer will start the HyperQueue server and create an allocation queue.  Currently all parameters related to the allocation queue are managed in the `hq_scripts/allocation_queue.sh` file, which can and likely should be changed before hosting the load balancer. You can find a detailed explanation of all parameters in the [HQ docs](https://it4innovations.github.io/hyperqueue/stable/deployment/allocation/#parameters), but the essential ones for our purposes are:
    - `--worker-start-command`: this specifies a command (`hq_scripts/setup_model.sh`) that will be run before a worker is started in a SLURM job. Its purpose is to setup the model server and to send back the URL to the load balancer (once again by writing to a file). At the moment it is required that all paths related to the `worker-start-cmd` are absolute because the command will run in a subdirectory of `.hq-server` and NOT in the cwd. This means that `hq_scripts/setup_model.sh` definitely needs to be adjusted before launching the load balancer.
    - `--resource`: here we define a new resource called "model" which will allow HQ to take the availability of a model into account for scheduling.
    - `-- ...`:  any parameters after `--` will be passed directly to `sbatch`. Some clusters require users to always specify certain SLURM parameters before submission (e.g. credentials, partition, etc.), so this should be set before attempting to run the load balancer. Also models that do actual computations may require specific configurations like number of gpus, certain node types, etc.
3. Whenever a request is received the load balancer will submit a new HQ job via `hq_scripts/job.sh`. Once again there are some configuration options, the most important one being that the job requests the resource "model" to get exclusive access. The HQ job then writes the ID of the SLURM job that it is running in into a file, which contains the HQ jobs ID in its name, and sleeps. The load balancer then reads two files, one to get the SLURM job ID and another to get the server URL of the model. Finally, the request can be processed as described in the beginning.

### How to run the load balancer
Short summary of the local changes that need to be made to do a test run:
- Provide a model that you want to run. As an example, you could compile a minimal C++ model, which simply multiplies the input by two, with the command `g++ -O3 -w -std=c++11 test/minimal/minimal-server.cpp -o server -lssl -lcrypto -pthread`.
- In `hq_scripts/setup_model.sh`: set the variables `load_balancer_dir` and `server_file` (paths need to be absolute!). Also, make sure that the script loads all dependencies and starts the model that you selected.
- In `hq_scripts/allocation_queue.sh`: add any SLURM arguments after `--` that are required to submit a job on your cluster.

### Problems and further work
The following is an (incomplete) list of problems that the current implementation has and contains some ideas for improvements.
- Performance issues: there is an immense overhead for every request (maybe around 5s?). There are most likely two components that cause this:
    -  Communication between jobs and the load balancer via the shared file system is very slow and not a very robust solution in general. There should be a better way to send the URL of the model server back to the load balancer. A possible idea could be to use HTTP for this as well, but I'm not sure how exactly this would work in practice. Alternatively, we could ask the HyperQueue developers to provide some sort of native solution (e.g. having a command that will wait for a job to get scheduled and then return the SLURM job ID).
    - Model functions like `Evaluate` will call other functions like `GetInputSizes` to verify the validity of the request (which is a good thing!). However, the problem is that the load balancer has to submit a new HQ job for every function call. This means that a single `Evaluate`request requires many jobs to be submitted, which can be very slow. It might be worthwile to somehow cache the results of basic functions like `SupportsEvaluate` or `GetInputSizes`, although this might not work for models that allow input sizes to be dependent on `config`. It might also turn out not to be much of a problem if we manage to make job submission more efficient.
- Streamlining usage/configuration: currently some local changes in various files are required for the load balancer to work. It might be desirable to move all configuration options into a single location to make the user experience smoother. Especially having to deal with absolute paths for the `worker-start-cmd` is very awkward and it would be great if we could find a better solution.
- Starting the load balancer: `test-LoadBalancer.cpp` always starts a manual SLURM job just to determine the names of the models. If jobs are very resource heavy and take a long time to get scheduled, this might be very inconvenient. Perhaps a user should simply manually specify model names. At the very least we should stop using manual SLURM jobs and fully switch to HQ instead.
- Fault tolerance: many unexpected things could happen in an environment that has this many components. A major improvement could be to ensure that all possible exceptions are handled in some way (e.g. what if the HyperQueue server is unable to start?) and that users receive appropriate feedback if a fatal error occurs.

Perhaps the entire approach of basically tricking HQ into doing the scheduling by submitting dummy jobs is not that great and there might be a much better solution.

Schlevidon committed:
Add basic HyperQueue integration and minimal test

Scripts require some manual configuration by a user.
Model address is sent to the load balancer via a shared filesystem.

linusseelinger commented on hq_scripts/job.sh:
```diff
@@ -0,0 +1,13 @@
+# and to ensure that HQ won't schedule any more jobs to this allocation
+
+echo "$SLURM_JOB_ID" > urls/hqjob-$HQ_JOB_ID.txt # send the slurm job id to load-balancer
+sleep 1h # keep the job occupied
```
We might as well go with an infinite loop of 1h sleeps here. Some models might take >1h to run

linusseelinger commented:
* We are now setting up the model instance when the worker is started via `--worker-start-command` , and I think that's where it should be. There could however be models that take quite some time during initial setup (i.e. in the constructor of their model class). Can we have some mechanism to ensure hq only sends tasks to the worker once the model is actually ready? Akin to readiness probes in kubernetes, we could for example see if the /Info endpoint of the model server is responsive. If so, we can be sure the model is ready for evaluations etc. Maybe that could just be a loop in the `--worker-start-command`.
If you don't see an obvious and reliable solution, we could merge your PR anyways and keep this as a separate task (so we can split work between you and Darius).
* Likewise, I think we can proceed with the issue of having an overhead due to input checks in the load balancer (thanks for spotting this!). I believe the cleanest solution would be to bypass all input/output checks in the loadbalancer, and pass through any errors to the client. Would need some work to do that cleanly though, so that can be a separate task.

Schlevidon committed:
Fix HQ job sleep timer

HQ jobs should now run forever to enable models that require more than one hour.

Schlevidon committed:
Merge branch 'main' into hyperqueue-integration
